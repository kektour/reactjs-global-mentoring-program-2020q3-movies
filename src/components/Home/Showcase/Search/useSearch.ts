import { useDispatch, useSelector } from 'react-redux';
import { moviesFilterSelector, MoviesSetSearchFilterAction, moviesSetSearchFilterActionCreator } from '../../../../store/movies';

type UseSearch = {
  value: string;
  setValue: (search: string) => MoviesSetSearchFilterAction;
};

export const useSearch = (): UseSearch => {
  const dispatch = useDispatch();
  const { search } = useSelector(moviesFilterSelector);
  const setSearchMovies = (search: string) => dispatch(moviesSetSearchFilterActionCreator(search));

  return {
    value: search,
    setValue: setSearchMovies,
  };
};
