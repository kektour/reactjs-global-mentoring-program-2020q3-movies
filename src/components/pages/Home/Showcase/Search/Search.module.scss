@import 'src/styles/shared.scss';

.root {
  display: flex;
}

.field {
  margin-right: 20px;
  border-radius: 5px;
  border: none;
  height: 60px;
  flex-grow: 1;
  font-size: 20px;
  background-color: rgba(85, 85, 85, 0.6);
  color: #fff;
  padding: 0 15px;

  &:focus {
    border: 1px solid #555555;
    outline: none;
  }
}

.button {
  @extend %button;
  width: 230px;
}
