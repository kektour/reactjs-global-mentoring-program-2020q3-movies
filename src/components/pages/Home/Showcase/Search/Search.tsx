import React, { useEffect } from 'react';
import { useFormik } from 'formik';
import { useRouter } from 'next/router';

import styles from './Search.module.scss';

export const Search: React.FC<{}> = () => {
  const router = useRouter();
  const formik = useFormik({
    initialValues: { search: router.query['query'] || '' },
    onSubmit: (values) => {
      if (values.search) {
        router.push(`/search/${values.search}`);
      } else {
        router.push('/');
      }
    },
    onReset: () => {
      formik.setValues({ search: router.query['query'] || '' });
    },
  });
  useEffect(() => {
    if (router.query['query']) {
      formik.setFieldValue('search', router.query['query']);
    }
  }, [router.query]);

  return (
    <form className={styles.root} onSubmit={formik.handleSubmit} onReset={formik.handleReset}>
      <input
        className={styles.field}
        type="text"
        name="search"
        value={formik.values.search}
        onChange={formik.handleChange}
        placeholder="What do you want to watch?"
      />
      <button className={styles.button} type="submit">
        Search
      </button>
    </form>
  );
};
