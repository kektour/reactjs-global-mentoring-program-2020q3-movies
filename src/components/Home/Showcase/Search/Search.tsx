import React from 'react';
import { useFormik } from 'formik';
import { useHistory } from 'react-router-dom';
import { useSearch } from './useSearch';

import styles from './Search.module.scss';

export const Search: React.FC<{}> = () => {
  const history = useHistory();
  const search = useSearch();
  const formik = useFormik({
    initialValues: { search: search.value },
    onSubmit: (values) => {
      if (values.search) {
        history.push(`/search/${values.search}`);
      } else {
        history.push('/');
      }
      search.setValue(values.search);
    },
    onReset: () => {
      formik.setValues({ search: search.value });
    },
  });

  return (
    <form className={styles.root} onSubmit={formik.handleSubmit} onReset={formik.handleReset}>
      <input
        className={styles.field}
        type="text"
        name="search"
        value={formik.values.search}
        onChange={formik.handleChange}
        placeholder="What do you want to watch?"
      />
      <button className={styles.button} type="submit">
        Search
      </button>
    </form>
  );
};
