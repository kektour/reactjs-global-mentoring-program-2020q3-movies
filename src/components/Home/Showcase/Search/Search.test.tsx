import React from 'react';
import { act, createEvent, fireEvent, render, wait } from '@testing-library/react';
import { Search } from './Search';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import userEvent from '@testing-library/user-event';
import * as hooks from './useSearch';

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const mockStore = configureStore();

describe('Search', () => {
  const store = mockStore({
    movies: {
      filter: {
        genre: 'foo',
        sortBy: 'bar',
        search: 'baz',
      },
    },
  });

  it('Should render component correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Search />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should user interact with the form', async () => {
    const { container, asFragment } = render(
      <Provider store={store}>
        <Search />
      </Provider>
    );
    await act(async () => await userEvent.type(container.querySelector('input[name=search]') as Element, 'Search Foo'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('User should submit form and be redirected', async () => {
    const searchValue = 'Search Foo';
    const setValueMock = jest.fn();
    const useSearchMock = jest.spyOn(hooks, 'useSearch').mockImplementation(() => ({
      value: (store.getState() as any).movies.filter.search,
      setValue: setValueMock,
    }));
    const { container } = render(
      <Provider store={store}>
        <Search />
      </Provider>
    );
    await act(async () => await userEvent.type(container.querySelector('input[name=search]') as Element, searchValue));
    act(() => userEvent.click(container.querySelector('button[type=submit]') as Element));

    await wait(() => {
      expect(setValueMock).toHaveBeenCalledWith(searchValue);
    });
    expect(mockHistoryPush).toHaveBeenCalledWith(`/search/${searchValue}`);

    useSearchMock.mockRestore();
  });

  it('User should submit form and be redirected â„–2', async () => {
    const setValueMock = jest.fn();
    const useSearchMock = jest.spyOn(hooks, 'useSearch').mockImplementation(() => ({
      value: (store.getState() as any).movies.filter.search,
      setValue: setValueMock,
    }));
    const { container } = render(
      <Provider store={store}>
        <Search />
      </Provider>
    );
    act(() => {
      fireEvent.change(container.querySelector('input[name=search]') as Element, {
        target: {
          value: '',
        },
      });
    });

    act(() => userEvent.click(container.querySelector('button[type=submit]') as Element));

    await wait(() => {
      expect(setValueMock).toHaveBeenCalledWith('');
    });
    expect(mockHistoryPush).toHaveBeenCalledWith('/');

    useSearchMock.mockRestore();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });
});
