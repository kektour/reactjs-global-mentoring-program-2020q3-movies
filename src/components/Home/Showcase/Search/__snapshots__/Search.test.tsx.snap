// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Should render component correctly 1`] = `
<DocumentFragment>
  <form
    class="root"
  >
    <input
      class="field"
      name="search"
      placeholder="What do you want to watch?"
      type="text"
      value="baz"
    />
    <button
      class="button"
      type="submit"
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;

exports[`Search Should user interact with the form 1`] = `
<DocumentFragment>
  <form
    class="root"
  >
    <input
      class="field"
      name="search"
      placeholder="What do you want to watch?"
      type="text"
      value="Search Foo"
    />
    <button
      class="button"
      type="submit"
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;
